origami image/operations

lmeiucsd opened the pull request:
SVG files are in assets/icons. I created placeholder descriptive naming conventions for the SVG file names for now.

lmeiucsd committed:
exported fortune teller states as SVG

phentos commented:
We don't need to merge this branch until the fortune teller is functional (which likely won't be for another week) so I'm converting the PR to draft, but I'll make an issue to clean up the SVGs

lmeiucsd committed:
cleaned up SVGs, renamed folder to 'origami'

lmeiucsd commented:
above commit is working on #20 as well as renaming the folder that stores the SVGs to 'origami'

phentos committed:
Merge pull request #27 from cse110-sp23-group1/main

sync with main

phentos committed:
pretty print the SVGs

phentos committed:
Merge pull request #34 from cse110-sp23-group1/main

sync from main

adarsh249 committed:
Updating Format of SVG so its easier to read

samw0627 committed:
Fixed svgs with IDs

adarsh249 committed:
Fixing my repo idk wtf happened lol

phentos committed:
add shell for svg handler as Origami.js

lmeiucsd committed:
added html and css to display SVG on web page

samw0627 committed:
Merge branch 'core-svg-handler' of https://github.com/cse110-sp23-group1/Origami-Fortune-Teller into core-svg-handler

samw0627 committed:
zipped files

adarsh249 committed:
Pushing new svg exports

adarsh249 committed:
Merge branch 'core-svg-handler' of https://github.com/cse110-sp23-group1/Origami-Fortune-Teller into core-svg-handler

adarsh249 committed:
Working on making clickable flaps... queryselector issues

lmeiucsd committed:
fixed queryselector issue, flaps respond to clicks

adarsh249 committed:
Created variable to understand which flap we have clicked.

adarsh249 committed:
Fixed up Origami class, cleaned up main.js, sidenote: gonna eat dinner then get back on to fix this up

adarsh249 committed:
Starting simple stop motin animation

adarsh249 committed:
Commit so i can work on my laptop

adarsh249 committed:
Got basic animations working

adarsh249 committed:
Basic functionality of SVG complete

phentos committed:
switch cases begone

samw0627 committed:
Updated Origami svg

Changed size for horizontally opened(no nums) and vertically opened(no nums). As well as added origami state with fortune

samw0627 committed:
Merge branch 'core-svg-handler' of https://github.com/cse110-sp23-group1/Origami-Fortune-Teller into core-svg-handler

adarsh249 committed:
Updated filepaths for new Figma SVGs

adarsh249 committed:
Centered fortune teller

lmeiucsd committed:
added documentation for Origami.js and main.js

phentos committed:
sync gitignore with main

phentos committed:
Delete .DS_Store

phentos committed:
Delete .DS_Store

phentos committed:
Delete .DS_Store

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+const horizontalNumsSVG = new Origami(svgPaths[1]);
+const verticalNumsSVG = new Origami(svgPaths[2]);
+const horizontalSVG = new Origami(svgPaths[3]);
+const verticalSVG = new Origami(svgPaths[4]);
```
Why are there multiple instances of `Origami` ? There is only one fortune teller, so there should only be one class. Move the state tracking of the svg into its state handler (the Origami class).

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+    if (CURRENTSVG !== closedSVG) {
+      CURRENTSVG.getFlapNumClicked().then((flapClicked) => {
+        const numAnimations = CURRENTSVG.getNumAnimations(flapClicked);
+        startAnimation(numAnimations);
```
It's hard to tell from the code, but it seems this is trying to tell what number the flap had in order to open and close the fortune teller that many times. If so, this approach seems very overengineered to me -- couldn't the click event call the animator and pass in the number directly?

Side note -- all of this belongs in the class itself

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+const verticalSVG = new Origami(svgPaths[4]);
+let animationCount = 0;
+closedSVG.generateSVG();
+let CURRENTSVG = closedSVG;
```
aren't we using `CAPITALIZED_NAMES` for constants? also, the class tracks / can track the current SVG, so there isn't a reason to have a global variable in main for it

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+ * @returns {Promise} - Promise object that animates between horizontally and vertically opened SVGs every 500ms, for a total of numAnimations - 1 times.
+ * After the animation, show the last Figma mockup.
+ */
+function startAnimation(numAnimations) {
```
This function belongs in the class, not in main

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+  return new Promise((resolve) => {
+    let animationIndex = 0;
+    animationCount++;
+    function animate() {
```
Give `animate` a proper function block, putting an async recursive function inside an anonymous function generating promises is brain-melting. For an example of how to more cleanly deal with promises, check out how i set them up for [the magic 8ball shakes](https://github.com/phentos/johnnymagic/blob/6c9208d2da89bca6dc0bd0072465be1ac25d24da/magic8ball.js#L239).

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+            break;
+          default:
+            // do nothing
+            break;
```
I'm allergic to `switch` in general, but to put it in more general terms why this isn't a good use case let's look at what it's doing.

- Given `CURRENTSVG`, check which type it is
- no matter what, call `removeCurrentSVG` on it (which is odd since "currentSVG" should already know what is current, right?)
- no matter what, set it to the opposite type

Since we're doing the same things no matter what, it isn't a `switch` at all. Instead consider moving this into a function that does the `no matter whats` -- ie, it swaps the current SVG style.

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+          } else {
+            verticalNumsSVG.generateSVG();
+            CURRENTSVG = verticalNumsSVG;
+          }
```
Similar situation to above -- we don't actually have any uncertainty about the next required state which would require a check. Always remove, always set the final type swapped.

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
 
-  editFortuneBox.style.display = 'block';
+/**
+ * Handles logic for the front-end animation of the fortune teller.
```
Since we're alternating between two states some number of times, we should just abstract the sequence. As in, given the animation is swapping `{horizontal, vertical}` however many times and then ending on some other state, why not set that up directly without any conditionals?

phentos commented on assets/scripts/main.js:
```diff
@@ -1,29 +1,92 @@
+const verticalNumsSVG = new Origami(svgPaths[2]);
+const horizontalSVG = new Origami(svgPaths[3]);
+const verticalSVG = new Origami(svgPaths[4]);
+let animationCount = 0;
```
Why is this a global? Just call the animation function with the demanded number, and it can handle it locally.

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+    'shadow-bottom-right-click': 4,
+    'shadow-top-left-click': 8,
+  },
+};
```
i converted these as a placeholder just to get them out of switch case, but this should be done cleaner; like couldn't we get the color from the click event itself? why do we need to look it up after the fact?

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+   * Creates an SVG object called "currentSVG" and adds click listeners
+   */
+  generateSVG() {
+    const svg = document.createElement('object');
```
why is this an `object` and not an `svg` ?

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+  generateSVG() {
+    const svg = document.createElement('object');
+    svg.data = this.svgPath;
+    document.body.appendChild(svg);
```
the class shouldn't be appending anything -- just give it back and let the caller decide where it goes

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+  removeCurrentSVG() {
+    this.currentSVG.remove();
+    this.currentSVG = null;
+  }
```
i'm not sure why we're removing the stored SVG at all -- why not just replace the paths that need to change to get to the next state?

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+      this.activateNumsHandler();
+    }
+  }
+  /**
+   * Function that converts ID of flap to its color.
+   * @param {string} - string that identifies which flap it is on the fortune teller
+   * @returns {number} - unknown if invalid input, or a string that represents color of flap
+   */
+  idToColor(flapID) {
+    return COLOR_BY_CLICK_REGION[flapID] || 'unknown';
+  }
```
Isn't the color of a flap in the SVG itself? Why do we need a lookup at all?

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+    return COLOR_BY_CLICK_REGION[flapID] || 'unknown';
+  }
+    
+  /**
+   * Function that converts ID of flap to its number specifically for SVGs with numbers.
+   * @param {string} - string that identifies which flap it is on the fortune teller
+   * @returns {number} - 0 if invalid input, or a number that is associated with the flap
+   */
+  idToNum(flapID) {
+    let number;
+
+    if (this.currentSVG.data.includes('horizontally')) {
+      number = NUMBER_BY_SHADOW_CLICK_REGION.horizontal[flapID];
+    } else if (this.currentSVG.data.includes('vertically')) {
+      number = NUMBER_BY_SHADOW_CLICK_REGION.vertical[flapID];
+    }
+
+    return (number !== undefined) ? number : 0;
+  }
```
Same as previous comment, the number is in the flap itself, so why do we need to look it up? I wrote about the number rendering [on slack](https://cse110team1.slack.com/archives/C0532PPASKE/p1685074449426429) a few days ago

phentos commented on assets/scripts/Origami.js:
```diff
@@ -0,0 +1,157 @@
+
+  /**
+   * Activates and handles click logic for the initial closed fortune teller SVG.
+   * @returns {Promise} - Promise object that adds a click event listener to each flap
+   */
+  activateClosedHandler() {
+    return new Promise((resolve) => {
+      this.currentSVG.addEventListener('load', () => {
+        const svgDocument = this.currentSVG.contentDocument;
+        const closedFlaps = svgDocument.querySelectorAll('g[id$="-flap-d"] path[id$="-click"]');
+        closedFlaps.forEach((flap) => {
+          flap.addEventListener('click', (event) => {
+            const flapID = flap.id;
+            const flapColorClicked = this.idToColor(flapID);
+            resolve(flapColorClicked);
+          });
+        });
+      });
+    });
+  }
+
+  /**
+   * Activates and handles click logic for the open fortune teller SVGs.
+   * @returns {Promise} - Promise object that adds a click event listener to each flap
+   */
+  activateNumsHandler() {
+    return new Promise((resolve) => {
+      this.currentSVG.addEventListener('load', () => {
+        const svgDocument = this.currentSVG.contentDocument;
+        const numFlaps = svgDocument.querySelectorAll('path[id$="-click"]');
+        numFlaps.forEach((flap) => {
+          flap.addEventListener('click', (event) => {
+            const flapID = flap.id;
+            const flapNumClicked = this.idToNum(flapID);
+            resolve(flapNumClicked);
+          });
+        });
+      });
+    });
+  }
```
these functions are identical aside from the query selector, but I'm fairly sure even that difference could be eliminated. These should be one function. I'm also not sure why this this needs to wait for load and be asynchronous? Why isn't it sufficient to just add the click listeners?

phentos reviewed (changes requested):
Flatten the SVG directories and get rid of the old versions -- there are currently almost 40 files, with almost all of them being unused.

We also should consider identifying the path changes between states and just apply those directly instead of hardcoding every svg variant

adarsh249 committed:
accidentally deleted wrong svg files lol, fixed now

adarsh249 committed:
Fixed style and filenames of SVGs

adarsh249 committed:
Cleaned up code a bit more, added all code to Origami.js, animation is still incredibly buggy, working on correcting this behavior

adarsh249 committed:
Created basic functionality of animations, small bugs with vertical/horizontal, need to still clean up code.

adarsh249 committed:
Removing text off Figma mockups

adarsh249 committed:
Fixed bug where user could click multiple times, added a darkening function and added this to a hover function on all clickable elements, saved Fortunes on first click to LocalStorage, removed sidebar when user starts animation, chose a random fortune from localStorage when user clicks final flap, adding files from sidebar branch to test svg core functionality.

adarsh249 committed:
Removed unncessary array

adarsh249 committed:
Added helper functions so if statements dont look too boggled

adarsh249 committed:
Fixed opening flap becoming smaller

lmeiucsd committed:
added docs for updated Origami.js

lmeiucsd committed:
Merge branch 'core-svg-handler' of https://github.com/cse110-sp23-group1/Origami-Fortune-Teller into core-svg-handler

lmeiucsd committed:
forgot to add file from git pull, this fixes size issues of SVGs

adarsh249 committed:
Working on getting text on the final flaps

adarsh249 committed:
small changes to start adding text on final flap

adarsh249 committed:
ReExported #-opened.svg to get text as <text> instead of <path>, also got text to appear on flap to work for only one flap

AlbertXDing committed:
Text works on every flap, fixed svg bug

AlbertXDing committed:
Fix InputFortineBox display bug when animations starts

lmeiucsd committed:
fixed svg bug that displays the wrong color when you hover your cursor over the inner flap while it's animating and then move your mouse out of it

lmeiucsd committed:
fixed hover bug for fully opened SVG

lmeiucsd committed:
delete console.log debug statements

lmeiucsd committed:
added and updated comments on my bug fix implementation

adarsh249 commented:
Please check the code now. Also, we put the sidebar on here for the status video and ended up writing some code about it anyways (hiding it when we animate, saving fortunes on first click, fetching fortunes from localStorage & displaying them at the end), so let us know if we need to remove that for this branch.

Salam-Aboul-Hosn committed:
Cursor Pointer styling for svg hover effect

parth4apple reviewed (commented):
Looks good! Just one small thing:
1. When you hover over the text of the number, you can not click the fortune flap to go to the next step. This can be fixed by putting `pointer-events: none` on the text element that contains the number.

adarsh249 committed:
reexported more svgs with text instead of path, created hover bug again

parth4apple reviewed (approved):
lgtm

lmeiucsd committed:
fixed hover bug on fully opened SVG

adarsh249 committed:
Resolved merge conflict from svg to main

adarsh249 committed:
lint fixing
